core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -207,6 +216,10 @@ public void uncaughtException(Thread t, Throwable e) {
   // switch to a ConcurrentHashMap.
   private final Set<InternalSubchannel> subchannels = new HashSet<>(16, .75f);
 
+  // Must be accessed from syncContext
+  @Nullable
+  private Collection<Runnable> pendingCalls = new ArrayList<>();
```

Comment:
Hmm, it might be hard to remove an item when it's cancelled. But at least we should avoid `setCall(newClientCall(method, callOptions))` when resolver returns after cancellation.

Reply:
DelayedClientTransport keeps a LinkedHashSet for the instances to be able to remove them. It doesn't seem that hard to me, although it'd be harder if you keep the collection containing Runnables.

I don't think `setCall(newClientCall(method, callOptions))` is any problem; ClientCallImpl can be safely thrown away and not started. And setCall() does nothing if the call was already.

Reply:
> DelayedClientTransport keeps a LinkedHashSet for the instances to be able to remove them. It doesn't seem that hard to me, although it'd be harder if you keep the collection containing Runnables.

I could use a `LinkedHashMap` instead. But I might need to copy the pattern from `DelayedClientTransport` as well.
```java
class PendingStream extends DelayedStream {
  @Override
  public void cancel(Status reason) {}
}
```

> I don't think setCall(newClientCall(method, callOptions)) is any problem

My concern was instantiation of `ClientCallImpl` objects just waiting for garbage collection.

Reply:
Doing removal for cancelled calls now.